### ISSUE TITLE
Unable to reset customer password after requiring province/state

### ISSUE THREAD

**Issue by teokolo** [magento/magento2#13179]
### Preconditions

1. Magento 2.1.11 clean install
2. PHP 5.6.33, Apache 2.2, Mysql-5.7.20
3. Single website, configured this way (probably bug is present for other combinations):
a) Default Country: Italy
b) Allowed Countries: Italy
c) State is Required for: do not set Italy (default setting: unselected)

### Steps to reproduce

1. Place an order registering an account, when asked for an address do not provide province/state
2. Logout from this customer account
3. Go to backend, set this option to require province/state:
- State is Required for: set Italy
4. Go to /customer/account/forgotpassword/ to reset password, insert customer email

### Expected result

1. Password reset link is sent 

### Actual result

1. Password reset link is not sent and an error message displays:

>  We\'re unable to send the password reset email.

var/log/system.log: 
```
==> magento/var/log/system.log <==
[2018-01-15 10:19:58] main.CRITICAL: Exception message: region is a required field.
Trace: #0 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/ResourceModel/AddressRepository.php(118): Magento\Customer\Model\ResourceModel\AddressRepository->_validate(Object(Magento\Customer\Model\Address))
#1 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/ResourceModel/CustomerRepository.php(343): Magento\Customer\Model\ResourceModel\AddressRepository->save(Object(Magento\Customer\Model\Data\Address))
#2 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/ResourceModel/CustomerRepository.php(197): Magento\Customer\Model\ResourceModel\CustomerRepository->updateAddresses(Object(Magento\Customer\Model\Data\Customer), '1')
#3 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(74): Magento\Customer\Model\ResourceModel\CustomerRepository->save(Object(Magento\Customer\Model\Data\Customer), NULL)
#4 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(70): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->___callParent('save', Array)
#5 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'save', Object(Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor), Array, 'update_newslett...')
#6 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-newsletter/Model/Plugin/CustomerPlugin.php(61): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Customer\Model\Data\Customer), NULL)
#7 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Newsletter\Model\Plugin\CustomerPlugin->aroundSave(Object(Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor), Object(Closure), Object(Magento\Customer\Model\Data\Customer), NULL)
#8 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'save', Object(Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor), Array, 'transactionWrap...')
#9 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/Plugin/CustomerRepository/TransactionWrapper.php(44): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Customer\Model\Data\Customer), NULL)
#10 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(142): Magento\Customer\Model\Plugin\CustomerRepository\TransactionWrapper->aroundSave(Object(Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor), Object(Closure), Object(Magento\Customer\Model\Data\Customer))
#11 /srv/www/test.mondoserviziweb.com/magento/var/generation/Magento/Customer/Model/ResourceModel/CustomerRepository/Interceptor.php(26): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->___callPlugins('save', Array, Array)
#12 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/AccountManagement.php(1213): Magento\Customer\Model\ResourceModel\CustomerRepository\Interceptor->save(Object(Magento\Customer\Model\Data\Customer))
#13 /srv/www/test.mondoserviziweb.com/magento/var/generation/Magento/Customer/Model/AccountManagement/Interceptor.php(271): Magento\Customer\Model\AccountManagement->changeResetPasswordLinkToken(Object(Magento\Customer\Model\Data\Customer), '42573c2086b9f14...')
#14 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/AccountManagement.php(513): Magento\Customer\Model\AccountManagement\Interceptor->changeResetPasswordLinkToken(Object(Magento\Customer\Model\Data\Customer), '42573c2086b9f14...')
#15 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(146): Magento\Customer\Model\AccountManagement->initiatePasswordReset('teo@matteogiann...', 'email_reset', NULL)
#16 /srv/www/test.mondoserviziweb.com/magento/var/generation/Magento/Customer/Model/AccountManagement/Interceptor.php(91): Magento\Customer\Model\AccountManagement\Interceptor->___callPlugins('initiatePasswor...', Array, Array)
#17 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Controller/Account/ForgotPasswordPost.php(73): Magento\Customer\Model\AccountManagement\Interceptor->initiatePasswordReset('teo@matteogiann...', 'email_reset')
#18 /srv/www/test.mondoserviziweb.com/magento/var/generation/Magento/Customer/Controller/Account/ForgotPasswordPost/Interceptor.php(24): Magento\Customer\Controller\Account\ForgotPasswordPost->execute()
#19 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/App/Action/Action.php(102): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->execute()
#20 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(74): Magento\Framework\App\Action\Action->dispatch(Object(Magento\Framework\App\Request\Http))
#21 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(70): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->___callParent('dispatch', Array)
#22 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Array, 'storeCheck')
#23 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-store/App/Action/Plugin/StoreCheck.php(44): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#24 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Store\App\Action\Plugin\StoreCheck->aroundDispatch(Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#25 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Array, 'contextPlugin')
#26 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-store/App/Action/Plugin/Context.php(106): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#27 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Store\App\Action\Plugin\Context->aroundDispatch(Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#28 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Array, 'customer-app-ac...')
#29 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Model/App/Action/ContextPlugin.php(61): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#30 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Customer\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#31 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Array, 'weee-app-action...')
#32 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-weee/Model/App/Action/ContextPlugin.php(112): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#33 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Weee\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#34 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Array, 'tax-app-action-...')
#35 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-tax/Model/App/Action/ContextPlugin.php(91): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#36 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Tax\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#37 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Custome...', 'dispatch', Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Array, 'customer_accoun...')
#38 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-customer/Controller/Plugin/Account.php(60): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#39 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(142): Magento\Customer\Controller\Plugin\Account->aroundDispatch(Object(Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#40 /srv/www/test.mondoserviziweb.com/magento/var/generation/Magento/Customer/Controller/Account/ForgotPasswordPost/Interceptor.php(39): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->___callPlugins('dispatch', Array, Array)
#41 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/App/FrontController.php(55): Magento\Customer\Controller\Account\ForgotPasswordPost\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#42 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(74): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#43 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(70): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#44 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'requestPreproce...')
#45 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php(94): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#46 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#47 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'install')
#48 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Module/Plugin/DbStatusValidator.php(69): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#49 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\Framework\Module\Plugin\DbStatusValidator->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#50 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(63): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'front-controlle...')
#51 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php(73): Magento\Framework\Interception\Chain\Chain->Magento\Framework\Interception\Chain\{closure}(Object(Magento\Framework\App\Request\Http))
#52 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Chain/Chain.php(67): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#53 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Interception\Chain\Chain->invokeNext('Magento\\Framewo...', 'dispatch', Object(Magento\Framework\App\FrontController\Interceptor), Array, 'front-controlle...')
#54 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/module-page-cache/Model/App/FrontController/VarnishPlugin.php(55): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#55 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/Interception/Interceptor.php(142): Magento\PageCache\Model\App\FrontController\VarnishPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#56 /srv/www/test.mondoserviziweb.com/magento/var/generation/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, Array)
#57 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/App/Http.php(135): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#58 /srv/www/test.mondoserviziweb.com/magento/vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#59 /srv/www/test.mondoserviziweb.com/magento/index.php(39): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))
#60 {main} [] []
```

This bug affects also checkout for customers who didn't set province/state, I'm going to open a separate bug for it.
I suggest also checking every aspect of this setting change because it could be a design error, not simply a bug.

**Comment by teokolo** [comment]
Same error when logging in customer account and trying to change password: /customer/account/edit/
Error message: region is a required field.

**Comment by magento-engcom-team** [comment]
@teokolo, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by magento-engcom-team** [comment]
@teokolo The issue is already fixed in 2.2.2 onward.
Closing ticket for now. Please reopen if you need any further assistance with this.

**Comment by Ctucker9233** [comment]
@magento-engcom-team This issue was originally reported against 2.1. If it's not fixed in 2.1, why is this closed? If you point me to the right commits, I can create a backport.

**Comment by majaode** [comment]
@Ctucker9233 Agree, we cannot expect all magento users to upgrade.
